swift/Samples/SampleApp/Sources/CrossFadeContainer.swift

```diff
@@ -0,0 +1,79 @@
+import Workflow
+import WorkflowUI
+
+
+struct CrossFadeScreen: Screen {
+    private var screenType: Any.Type
+    var screen: AnyScreen
+    var key: String
+
+    init<ScreenType: Screen>(_ screen: ScreenType, key: String = "") {
+        self.screenType = ScreenType.self
+
+        self.screen = AnyScreen(screen)
+        self.key = key
+    }
+
+    fileprivate func isEquivalent(to otherScreen: CrossFadeScreen) -> Bool {
+        return (self.screenType == otherScreen.screenType) && (self.key == otherScreen.key)
```

Comment:
I could consider that, as long as there is an implicit key if it's not specified (that falls back to the screen type). Basically want it to always cross fade between different conceptual screens.

Otherwise, it requires the key to be specified with different screen types to get the animation, which is likely not ever what we would want.

Reply:
Updated this to only compare screen type *if* both keys are empty. That supports the naive case (where there isn't a key specified, but I want it to animate), but also gives the flexibility to switch two different screen types without it animating.

Reply:
Ah, yeah, good call. Kinda missed the default of no key.

Could be stored as an `AnyHashable` like:
```swift

init<ScreenType: Screen>(base screen: ScreenType, key: String? = nil) {
    self.baseScreen = AnyScreen(screen)
    if let key = key {
        self.key = AnyHashable(key)
    } else {
        self.key = AnyHashable(ObjectIdentifier(ScreenType.self))
    }
}
```

(this is mostly just academic in this case, but…)

Reply:
Might as well go all the way:

```swift
init<ScreenType: Screen, Key: Hashable>(base screen: ScreenType, key: Key? = nil) {
    self.baseScreen = AnyScreen(screen)
    if let key = key {
        self.key = AnyHashable(key)
    } else {
        self.key = AnyHashable(ObjectIdentifier(ScreenType.self))
    }
}
```

Because now I can do:

```swift

enum Key: Hashable {
    case loggedIn
    case loggedOut
}

return CrossFadeScreen(
    base: MyScreen(),
    key: .loggedIn)

```

I've begun to strongly favor generic key types (which become `AnyHashable`) over plain strings.

Reply:
I'm good with using `AnyHashable` instead of a string over all.

However, at least with XCode 10.2, we would need an extra initializer for the case without a key - otherwise calling `CrossFadeScreen(base: myScreen)` will error with "Generic parameter 'Key' could not be inferred".
```
struct CrossFadeScreen: Screen {
    var baseScreen: AnyScreen
    var key: AnyHashable

    init<ScreenType: Screen, Key: Hashable>(base screen: ScreenType, key: Key?) {
        self.screenType = ScreenType.self

        self.baseScreen = AnyScreen(screen)
        if let key = key {
            self.key = AnyHashable(key)
        } else {
            self.key = AnyHashable(ObjectIdentifier(ScreenType.self))
        }
    }

    init<ScreenType: Screen>(base screen: ScreenType) {
        let key = Optional<AnyHashable>.none
        self.init(base: screen, key: key)
    }

    fileprivate func isEquivalent(to otherScreen: CrossFadeScreen) -> Bool {
        return self.key == otherScreen.key
    }
}
```

Updated based on that.